File: DAA/Week1/3.py
import random
import time

def insertion_sort(arr, n):
    if n<=1:
        return
    insertion_sort(arr,n-1)
    last = arr[n-1]
    j = n-2
    while (j>=0 and arr[j]>last):
        arr[j+1] = arr[j]
        j = j-1
    arr[j+1]=last


n = 5000
arr = []
print("Input size:", n)

# Best Case: Array is already sorted
for i in range(n):
    arr.append(i) 
start = time.time()
insertion_sort(arr,n)
end = time.time()
print("Best Case:", '{:.4f}'.format(end - start), "seconds")
arr.clear()

# Average Case: Array is random
for i in range(n):
    arr.append(random.randint(0, n))
start = time.time()
insertion_sort(arr,n)
end = time.time()
print("Average Case:", '{:.4f}'.format(end - start), "seconds")
arr.clear()

# Worst Case: Array is sorted in reverse order
for i in range(n):
    arr.append(n-i)
start = time.time()
insertion_sort(arr,n)
end = time.time()
print("Worst Case:", '{:.4f}'.format(end - start), "seconds")
arr.clear()

File: DAA/C Implementation/Week1/3.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void insertion_sort(int arr[], int n)
{
    if (n <= 1)
        return;
    insertion_sort(arr, n - 1);
    int last = arr[n - 1];
    int j = n - 2;
    while (j >= 0 && arr[j] > last) {
        arr[j + 1] = arr[j];
        j--;
    }
    arr[j + 1] = last;
}

int main()
{
    int n = 10000, i;
    clock_t start, end;
    int *arr;

    // Allocating memory dynamically
    arr = (int *)malloc(n * sizeof(int));
    if (arr == NULL)
    {
        printf("Memory allocation failed\n");
        return 0;
    }

    // Best Case: Array is already sorted
    for (i = 0; i < n; i++)
        arr[i] = i;
    start = clock();
    insertion_sort(arr, n);
    end = clock();
    printf("Best Case: %.2f ms\n", ((double)(end - start)) * 1000 / CLOCKS_PER_SEC);

    // Average Case: Array is random
    for (i = 0; i < n; i++)
        arr[i] = rand() % n;
    start = clock();
    insertion_sort(arr, n);
    end = clock();
    printf("Average Case: %.2f ms\n", ((double)(end - start)) * 1000 / CLOCKS_PER_SEC);

    // Worst Case: Array is sorted in reverse order
    for (i = 0; i < n; i++)
        arr[i] = n - i;
    start = clock();
    insertion_sort(arr, n);
    end = clock();
    printf("Worst Case: %.2f ms\n", ((double)(end - start)) * 1000 / CLOCKS_PER_SEC);

    free(arr);
    return 0;
}

